core/controllers/question_test.py

```diff
@@ -111,3 +111,350 @@ def test_get(self):
+        question_services.add_question_id_to_skill(
+            question_id, self.collection_id,
+            self.skill_id, self.new_user_id)
+
+        collection_services.record_played_exploration_in_collection_context(
+            self.new_user_id, self.collection_id, self.exp_id)
+        change_list = [{'cmd': 'update_question_property',
+                        'property_name': 'title',
+                        'new_value': 'ABC',
+                        'old_value': 'A Question'}]
+        self.payload['change_list'] = json.dumps(change_list)
+        self.payload['commit_message'] = 'update title'
+        self.login(self.NEW_USER_EMAIL)
+        response_json = self.put_json(
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id, question_id),
+            self.payload, expect_errors=False)
+        self.assertIn('question_id', response_json.keys())
+
+        del self.payload['change_list']
+        self.put_json(
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id,
+                question_id), self.payload, expect_errors=True,
+            expected_status_int=404)
+
+        del self.payload['commit_message']
+        self.payload['change_list'] = json.dumps(change_list)
+        self.put_json(
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id,
+                question_id), self.payload, expect_errors=True,
+            expected_status_int=404)
+
+        self.payload['commit_message'] = 'update title'
+        self.put_json(
+            '%s/%s' % (feconf.QUESTION_DATA_URL, self.collection_id),
+            self.payload, expect_errors=True,
+            expected_status_int=404)
+
+    def test_delete(self):
+        question = question_domain.Question(
+            'dummy', 'A Question',
+            exp_domain.State.create_default_state('ABC').to_dict(), 1,
+            self.collection_id, 'en')
+
+        question_id = question_services.add_question(
+            self.new_user_id, question)
+        question_services.add_question_id_to_skill(
+            question_id, self.collection_id, self.skill_id,
+            self.new_user_id)
+        self.login(self.NEW_USER_EMAIL)
+        response = self.testapp.delete(
+            '%s/%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id, question_id),
+            expect_errors=False)
+        self.assertEqual(response.status_int, 200)
+
+        response = self.testapp.delete(
+            '%s/%s' % (
+                feconf.QUESTION_DATA_URL, self.collection_id),
+            expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+        response = self.testapp.delete(
+            '%s/' % feconf.QUESTION_DATA_URL, expect_errors=True)
+        self.assertEqual(response.status_int, 404)
+
+
+class QuestionsPostHandlerTest(test_utils.GenericTestBase):
+    """Test the question handler."""
+
+    def setUp(self):
```

Comment:
If the setUp is the same as the above class then why not just combine the two classes? You can have more than one test method in a class. For each test method, setUp() will be run before that test, and all test methods are run independently.

Ditto below. In general try not to duplicate lots of code.

Reply:
Updated.